src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/ConvertToRecordHelpers.cs

```diff
@@ -0,0 +1,914 @@
+        /// <param name="builder">builder for property/field comparisons we encounter</param>
+        /// <returns>whether every statment was one of the expected forms</returns>
+        private static bool TryAddEqualizedFieldsForStatements(
+            IEnumerable<IOperation> statementsToCheck,
+            ISymbol otherC,
+            INamedTypeSymbol type,
+            ArrayBuilder<IFieldSymbol> builder)
+            => statementsToCheck.FirstOrDefault() switch
+            {
+                IReturnOperation
+                {
+                    ReturnedValue: ILiteralOperation
+                    {
+                        ConstantValue.HasValue: true,
+                        ConstantValue.Value: true,
+                    }
+                }
+                    // we are done with the comparison, the final statment does no checks
+                    => true,
+                IReturnOperation { ReturnedValue: IOperation value } => TryAddEqualizedFieldsForCondition(
+                    value, successRequirement: true, currentObject: type, otherObject: otherC, builder: builder),
+                IConditionalOperation
+                {
+                    Condition: IOperation condition,
+                    WhenTrue: IOperation whenTrue,
+                    WhenFalse: var whenFalse,
+                }
+                    // 1. Check structure of if statment, get success requirement
+                    // and any potential statments in the non failure block
+                    // 2. Check condition for compared members
+                    // 3. Check remaining members in non failure block
+                    => TryGetSuccessCondition(whenTrue, whenFalse, statementsToCheck.Skip(1).AsImmutable(),
+                        out var successRequirement, out var remainingStatements) &&
+                    TryAddEqualizedFieldsForCondition(
+                            condition, successRequirement, type, otherC, builder) &&
+                    TryAddEqualizedFieldsForStatements(remainingStatements, otherC, type, builder),
+                _ => false
+            };
+
+        private static bool TryAddFieldFromComparison(
+            IMemberReferenceOperation memberReference1,
+            IMemberReferenceOperation memberReference2,
+            ISymbol currentObject,
+            ISymbol otherObject,
+            ArrayBuilder<IFieldSymbol> builder)
+        {
+            var leftObject = GetReferencedSymbolObject(memberReference1.Instance!);
+            var rightObject = GetReferencedSymbolObject(memberReference2.Instance!);
+            if (memberReference1.Member.Equals(memberReference2.Member) &&
+                leftObject != null &&
+                rightObject != null &&
+                !leftObject.Equals(rightObject) &&
+                (leftObject.Equals(currentObject) || leftObject.Equals(otherObject)) &&
+                (rightObject.Equals(currentObject) || rightObject.Equals(otherObject)))
```

Comment:
Given we've had this pattern enough in this file I wonder if it's worth having a "these two objects are those two objects, in any order" helper.